esutil/bulk_indexer.go

```diff
@@ -0,0 +1,535 @@
+	Index           string
+	Action          string
+	DocumentID      string
+	Body            io.Reader
+	RetryOnConflict *int
+
+	Metadata interface{} // To use eg. in OnSuccess/OnFailure callbacks
+
+	OnSuccess func(item BulkIndexerItem, res BulkIndexerResponseItem)            // Per item
+	OnFailure func(item BulkIndexerItem, res BulkIndexerResponseItem, err error) // Per item
+}
+
+// BulkIndexerResponse represents the Elasticsearch response.
+//
+type BulkIndexerResponse struct {
+	Took      int                                  `json:"took"`
+	HasErrors bool                                 `json:"errors"`
+	Items     []map[string]BulkIndexerResponseItem `json:"items,omitempty"`
+}
+
+// BulkIndexerResponseItem represents the Elasticsearch response item.
+//
+type BulkIndexerResponseItem struct {
+	Index      string `json:"_index"`
+	DocumentID string `json:"_id"`
+	Version    int64  `json:"_version,omitempty"`
+	Result     string `json:"result,omitempty"`
+	Status     int    `json:"status,omitempty"`
+	SeqNo      int64  `json:"_seq_no,omitempty"`
+	PrimTerm   int64  `json:"_primary_term,omitempty"`
+
+	Shards struct {
+		Total      int `json:"total,omitempty"`
+		Successful int `json:"successful,omitempty"`
+		Failed     int `json:"failed,omitempty"`
+	} `json:"_shards,omitempty"`
+
+	Error struct {
+		Type   string `json:"type"`
+		Reason string `json:"reason"`
+		Cause  struct {
+			Type   string `json:"type"`
+			Reason string `json:"reason"`
+		} `json:"caused_by"`
+	} `json:"error,omitempty"`
+}
+
+// BulkResponseJSONDecoder defines the interface for custom JSON decoders.
+//
+type BulkResponseJSONDecoder interface {
+	UnmarshalFromReader(io.Reader, *BulkIndexerResponse) error
+}
+
+// BulkIndexerDebugLogger defines the interface for a debugging logger.
+//
+type BulkIndexerDebugLogger interface {
+	Printf(string, ...interface{})
+}
+
+type bulkIndexer struct {
+	wg      sync.WaitGroup
+	queue   chan BulkIndexerItem
+	workers []*worker
+	ticker  *time.Ticker
+	stats   *bulkIndexerStats
+
+	config BulkIndexerConfig
+}
+
+type bulkIndexerStats struct {
+	numAdded    uint64
+	numFlushed  uint64
+	numFailed   uint64
+	numIndexed  uint64
+	numCreated  uint64
+	numUpdated  uint64
+	numDeleted  uint64
+	numRequests uint64
+}
+
+// NewBulkIndexer creates a new bulk indexer.
+//
+func NewBulkIndexer(cfg BulkIndexerConfig) (BulkIndexer, error) {
+	if cfg.Client == nil {
+		cfg.Client, _ = elasticsearch.NewDefaultClient()
+	}
+
+	if cfg.Decoder == nil {
+		cfg.Decoder = defaultJSONDecoder{}
+	}
+
+	if cfg.NumWorkers == 0 {
+		cfg.NumWorkers = runtime.NumCPU()
+	}
+
+	if cfg.FlushBytes == 0 {
+		cfg.FlushBytes = 5e+6
+	}
+
+	if cfg.FlushInterval == 0 {
+		cfg.FlushInterval = 10 * time.Second
+	}
+
+	bi := bulkIndexer{
+		config: cfg,
+		stats:  &bulkIndexerStats{},
+	}
+
+	bi.init()
+
+	return &bi, nil
+}
+
+// Add adds an item to the indexer.
+//
+func (bi *bulkIndexer) Add(ctx context.Context, item BulkIndexerItem) error {
```

Comment:
@axw, one thing I haven't yet figured out is how to properly instrument the underlying client with this interface. It's nice that `Add()` takes a context, but there's no way to propagate that to the `flush()` and `BulkRequest.Do()` down the line at the moment. One solution is to add an unexported field to `BulkIndexerItem` — the way the client does it —, and set it here, in this method. Do you have other ideas / suggestions?

Reply:
@karmi Elastic APM doesn't currently support assigning multiple parents to a transaction or span, so it wouldn't really work to propagate trace context through Add.

If you want to trace the bulk requests, I'd suggest instead creating transactions in the flusher. I'm beginning to think there should be a Flusher interface on the config, with a default implementation exported.

Reply:
Instrumentation support added in https://github.com/elastic/go-elasticsearch/commit/aa612f9daef5f04549ab8eca251cdf20eeb755f7, example available in:

https://github.com/elastic/go-elasticsearch/blob/0258f7a7f16791c3eaed36307f784efd9639b346/_examples/bulk/kafka/kafka.go#L142-L148